Introduce CI, and fix tests

meshy opened the pull request:
This introduces CI, and gets the tests running on Django 1.8-2.1.

Upstream are not able to continue supporting this library, so we're going to maintain our own fork.

It's well worth reviewing this one commit at a time.

When reviewing, please remember that this is a **public repo**.

This was previously split into two upstream PRs. One to get the CI up and running (https://github.com/talpor/django-forkit/pull/2) and one to get the tests passing (https://github.com/talpor/django-forkit/pull/3/).

I'll open another PR soon to get this ready for release as a new library.

meshy committed:
Replace distutils with setuptools

meshy committed:
Add test runner

This acts as an entry point for running the tests. To run them, call:

    python forkit/tests/run.py

This runs the tests against a Postgres database. This can be overridden
with a DATABASE_URL environment variable, as documented on the
django-environ website [0].

If you use the default DB settings, you'll need to create a database
before running the tests:

    psql -c "CREATE DATABASE forkit;"

To get the new Django test runner to discover the existing tests, I had
to rename the files so that they started with `test_`.

This also removes the custom test runner, colourer, and coverage module.
I'll restore this functionality in another commit.

[0]: https://django-environ.readthedocs.io/en/latest/

meshy committed:
Add tox, and instructions for running tests

It should be noted that every single combination of Django and Python
now listed in tox fails to run successfully. I'll attempt to address
that separately.

meshy committed:
Add code coverage reporting to tox

meshy committed:
Add a TravisCI config

meshy committed:
Fix strange import order in test runner

I wrote it this way when I was trying to get tests running against
Django < 1.8, but should have corrected it when I abandoned that idea.

meshy committed:
Allow new ways of running tests

This adds four new ways to alter the test-run. You can:

- run the tests in reverse order with `--reverse`.
- tell the tests to abort on first failure with `--failfast`.
- define how verbose you would like the tests to be with `--verbose`.
- run only specific tests by providing their dotted paths.

For example, to use all of them at once, you might run:

    python forkit/tests/run.py -v2 --reverse --failfast forkit.tests.test_utils

meshy committed:
Stop state from leaking out of signals tests

The call to `disconnect` didn't match the call to `connect`, so the
signals were not detached properly. Correcting this fixes other tests
that were failing as a result of the unexpectedly connected signal
handlers.

meshy committed:
Don't rely on exact primary keys in tests

Depending on the order that the tests are run in, the results can
differ.

meshy committed:
Don't save relations before related object

A row must be in the database before one can reference it with a
foreign-key.

https://docs.djangoproject.com/en/2.1/releases/1.8/#assigning-unsaved-objects-to-relations-raises-an-error

meshy committed:
Replace use of deprecated model _meta methods

Both get_all_related_many_to_many_objects and get_all_related_objects
were removed from Django 1.11 [0].

I followed Django's migration guide [1] to update them.

[0]: https://docs.djangoproject.com/en/2.1/releases/1.8/#selected-methods-in-django-db-models-options-options
[1]: https://docs.djangoproject.com/en/2.1/ref/models/meta/#migrating-old-meta-api

meshy committed:
Fix Python 3 syntax error in test runner

meshy committed:
Fix Python 3 issues to get more tests running

meshy committed:
Pass arguments to the test runner through tox

https://tox.readthedocs.io/en/latest/config.html#substitutions-for-positional-arguments-in-commands

meshy committed:
Ensure related objects are saved on Python 3

On Python 3, calls to map() are lazy, so replacing it with a for loop
ensures that the list of related objects is iterated.

meshy committed:
Use set() to populate ManyToMany relationships

It's available from Django 1.9 onwards. What's more, direct assignment to
ManyToMany relationships is prohibited from Django 2.0 onwards.

meshy committed:
Explicitly define "on_delete" on ForeignKeys

This is required in Django 2.0. I've set them to CASCADE because that
was the old default.

meshy committed:
Remove semicolon at EOL!

LilyFirefly commented on forkit/tests/test_fork.py:
```diff
@@ -18,7 +18,7 @@ def test_shallow_fork(self):
         fork = self.author.fork()
 
-        self.assertEqual(fork.pk, 3)
+        self.assertTrue(fork.pk > self.author.pk)
```
```python
self.assertGreater(fork.pk, self.author.pk)
```

LilyFirefly commented on forkit/tests/test_fork.py:
```diff
@@ -24,7 +24,7 @@ def test_shallow_fork(self):
 
         self.assertEqual(fork2.pk, None)
-        self.assertEqual(fork2._commits.related.keys(), ['posts'])
+        self.assertEqual(list(fork2._commits.related.keys()), ['posts'])
```
You could do:
```python
self.assertSequenceEqual(fork2._commits.related.keys(), ['posts'])
```

meshy replied:
Python 3 doesn't like this.

```pytb
======================================================================
FAIL: test_shallow_fork (forkit.tests.test_fork.ForkModelObjectTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charlie/code/virtualstock/django-forkit/forkit/tests/test_fork.py", line 27, in test_shallow_fork
    self.assertSequenceEqual(fork2._commits.related.keys(), ['posts'])
AssertionError: Sequences differ: dict_keys(['posts']) != ['posts']

Unable to index element 0 of first sequence

- dict_keys(['posts'])
+ ['posts']


```

meshy replied:
This works:

```diff
-        self.assertEqual(list(fork2._commits.related.keys()), ['posts'])
+        if not hasattr(self, 'assertCountEqual'):
+            # Python 2 compat
+            self.assertItemsEqual(fork2._commits.related.keys(), ['posts'])
+        else:
+            # Python 3
+            self.assertCountEqual(fork2._commits.related.keys(), ['posts'])
```

It's look-before-you-leap, rather than a try/except, but I think I prefer it this way, in this case, as it will make it easier to remove when Python 2 is dropped.

meshy replied:
Better yet:

```diff
-        self.assertEqual(list(fork2._commits.related.keys()), ['posts'])
+
+        if hasattr(self, 'assertCountEqual'):
+            # Python 3
+            assert_same_items = self.assertCountEqual
+        else:
+            # Python 2 compat
+            assert_same_items = self.assertItemsEqual
+        assert_same_items(fork2._commits.related.keys(), ['posts'])
```

meshy replied:
Clearly I'm not very decisive today. I've gone with a `try` after all.

LilyFirefly commented on forkit/commit.py:
```diff
@@ -34,7 +34,10 @@ def _commit_related(instance, memo, stack, **kwargs):
+            if hasattr(getattr(instance, accessor), 'set'):
+                getattr(instance, accessor).set(value)
+            else:
+                setattr(instance, accessor, value)
```
I'd prefer to write this as:

```python
relation = getattr(instance, accessor)
try:
    relation.set(value)
except AttributeError:
    setattr(instance, accessor, value)
```

It avoids calling `getattr` twice and looking for `set` twice. I think it's also a bit clearer to read.

rklyne replied:
I actually prefer the former because it is explicitly saying which attribute can be in error. We don't want to trap errors coming out of the set method 

That said, I don't feel too strongly about it.

meshy replied:
> it is explicitly saying which attribute can be in error

Ah, good point. I think I should be able to accommodate both requirements. 
:)

LilyFirefly commented on forkit/tests/run.py:
```diff
@@ -0,0 +1,38 @@
+parser.add_argument('tests', type=six.text_type, nargs='*', default=[])
+parser.add_argument('-v', '--verbosity', type=int, default=1)
+parser.add_argument('--failfast', action='store_true')
+parser.add_argument('--reverse', action='store_true')
```
Why not `--keepdb`? This is a reason I prefer just creating a dummy project and using `manage.py test` - you don't need to faff about reimplementing functionality built in to  `manage.py test`.

meshy replied:
Adding `keepdb` is a good idea. I'm not sold on adding a whole project, `manage.py`, and the rest though.

meshy replied:
Although I must admit I can see the appeal... Perhaps I should give it a shot to see how it compares.

meshy replied:
I've added `keepdb`, but having a test project will require a larger change, and I'm keen to see this merged. I'm open to considering adding it in future, but I'd like to leave it for the moment.

veryhappythings replied:
I'm not well versed in django library test practices, but it seems to me that whacking pytest in here would be beneficial. Is that a done thing? It feels strange to be implementing arguments on a per-project basis.

LilyFirefly commented on forkit/tests/models.py:
```diff
@@ -1,67 +1,79 @@
-    def __unicode__(self):
+    def __str__(self):
         return u'{0}'.format(self.title)
-
```
Are you going to add migrations for these test models?

meshy replied:
Ah, it hadn't occurred to me, as the tests ran without them. Is there an advantage to adding them?

meshy committed:
Use assertGreater where appropriate in tests

meshy committed:
Use try/except instead of repeating getattr

meshy committed:
Use assertCountEqual instead of casting to list

Python 2 calls assertCountEqual "assertItemsEqual", so we provide
backwards compatibility for that.

meshy committed:
Add keepdb flag to test runner

veryhappythings reviewed (approved):
Well the tests pass so it must be good! :D 

This looks good to me. One comment below, mainly to aid my understanding of how this stuff works...

meshy commented:
@veryhappythings I'll summarise our IRL discussion here, for posterity:

There isn't a standard  or best way to get the tests running, as far as I'm aware. It's something of a sticking point sometimes. While I'm not keen on this solution, I prefer it to any alternatives I'm familiar with. It's probably worth revisiting, but it's not a priority right now.

Also, for some reason I can't comment directly on your comment!

meshy commented:
I hope you don't mind my dismissing your review, @Ian-Foote. I think I've addressed your points, with the exception of the test project, which can be tackled at a later date.
